File: euler808/euler808.py
from math import sqrt



def is_square(N):
    r = round(sqrt(N))
    return r * r == N


def is_prime(n):
    for d in range(2, int(sqrt(n))):
        if n%d == 0:
            return False
    return True


found = []

for n in range(1, 50_000_000, 2):
    # only loop over perfect squares
    N = n*n

    if N % 10 not in [1,9]: continue

    s = str(N)

    if s[0] not in '19': continue    

    N_rev = int(s[::-1])

    # eliminate palindromes
    if N == N_rev: continue

    # eliminate if reverse is not also a perfect square
    if not is_square(N_rev): continue

    n_rev = int(sqrt(N_rev))

    # do primality check as late as possible
    if is_prime(n) and is_prime(n_rev):
        print(N)
        found.append(N)

        if len(found) == 50: break

print(sum(found))

# 3807504276997394

File: euler808/euler808.cpp
#include <cmath>
#include <string>
#include <iostream>
#include <algorithm>


bool is_square(long long N) {
    long long r = round(sqrt(N));
    return r*r == N;
}

bool is_prime(long long n) {
    if (n%2 == 0) return false;
    for (long long d = 3; d < sqrt(n); d+=2) {
        if (n%d == 0) {
            return false;
        }
    }

    return true;
}


long long reverse(long long n) {
    std::string s = std::to_string(n);
    std::reverse(s.begin(), s.end());
    return std::stoll(s);
}


void euler808() {
    long long total = 0;
    int num_found = 0;
    for (long long n = 1; num_found < 50; n+=2) {
        long long N = n*n;

        if (N%10 != 1 && N%10 != 9) continue;

        long long N_rev = reverse(N);

        if (N == N_rev) continue;

        if (!is_square(N_rev)) continue;

        long long n_rev =  round(sqrt(N_rev));

        if (is_prime(n) && is_prime(n_rev)) {
            std::cout << N << std::endl;

            num_found++;
            total += N;
        }
    }

    std::cout << total << std::endl;
}

#include "../helpers/timing.cpp"
int main() {
    auto duration = func_time_ms(euler808);
    std::cout << "\n(" << duration << " ms)" << std::endl;
}
